### PR TITLE
refactor!: update api.context, remove `entry`, `targets` properties

### PR DESCRIPTION
**Pull request by 9aoy**
## Summary

Remove `api.context.entry` and `api.context.targets`,  use `api.environment` instead.

 The `distPath` and `tsconfigPath` may be different in different environments. Rsbuild will try to get the parent distPath of environments as `context.distPath`, it is recommended to use `api.environment[name]` instead of `api.context`.

```diff
const pluginFoo = () => ({
  setup(api) {
      api.modifyBundlerChain((chain, { environment }) => {
-            const { entry,  targets, tsconfigPath, distPath } = api.context;
+            const { entry,  target, tsconfigPath, distPath } = api.environments[environment];
    });
  },
});

```

## Related Links

https://github.com/web-infra-dev/rsbuild/issues/2620

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/plugins/cache.ts [hunk 1]
```diff
@@ -126,8 +132,8 @@ export const pluginCache = (): RsbuildPlugin => ({
         // The default cache name of webpack is '${name}-${env}', and the `name` is `default` by default.
         // We set cache name to avoid cache conflicts of different targets.
         name: useDigest
-          ? `${target}-${env}-${getDigestHash(cacheConfig.cacheDigest!)}`
-          : `${target}-${env}`,
+          ? `${target}-${env}-${environment}-${getDigestHash(cacheConfig.cacheDigest!)}`
```
Comment (chenjiahan) [inline packages/core/src/plugins/cache.ts:135]: Maybe `target` can be removed

File: packages/core/src/createContext.ts [hunk 2]
```diff
@@ -165,11 +162,14 @@ export function updateContextByNormalizedConfig(
...
 ) {
-  context.distPath = getAbsoluteDistPath(context.rootPath, config);
-
-  if (config.source.entry) {
-    context.entry = getEntryObject(config, 'web');
-  }
+  // Try to get the root distPath of context.environments as context.distPath
+  context.distPath = Object.values(context.environments).reduce((p, e) => {
+    // foo/dist/server vs foo/dist
+    if (!p || p.startsWith(e.distPath)) {
```
Comment (chenjiahan) [inline packages/core/src/createContext.ts:168]: Can this function get the common parent directory as expected?

For example, `/path/to/dist/client` and `/path/to/dist/server`, the parent directory is `/path/to/dist`

Comment (9aoy) [reply packages/core/src/createContext.ts:168]: but if `/path/dist-client` and `/path/dist-server`, the parent directory is `/path`, is expected?

Comment (chenjiahan) [reply packages/core/src/createContext.ts:168]: Yes, because we can't use `/path/dist-client' or `/path/dist-server'.

**Review by chenjiahan — APPROVED** [review]

**Comment by colinaaa** [comment]
I cannot find `api.environments`(or `api.context.environments`) in Rsbuild `1.0.0-alpha`.

Seems like it has been changed to `utils.environments` in https://github.com/web-infra-dev/rsbuild/pull/2721.

But the `environements` does not exist in the utils of `api.onDevCompileDone`.

So, how could I migrate the following code:

```js
api.onDevCompileDone(async ({ isFirstCompile, stats }) => {
  const entries = Object.keys(api.context.entry)
})
```

**Comment by chenjiahan** [comment]
Yeah we have changed the design and no longer provides "api.context.environments".

The onDevCompileDone callback should have a environements param, @9aoy do we missing it

**Comment by 9aoy** [comment]
> So, how could I migrate the following code:
> 
> ```js
> api.onDevCompileDone(async ({ isFirstCompile, stats }) => {
>   const entries = Object.keys(api.context.entry)
> })
> ```

@colinaaa  The environment api is not fully complete yet (also doc and features). And it seems that the `environments` parameter is missing in `onDevCompileDone` hook.
When completed, it should be migrated as follows:

```
api.onDevCompileDone(async ({ isFirstCompile, stats, environments }) => {
  const entries = Object.keys(environments).map(e => e.entry)
})
```

**Pull request merged**